stl/inc/xutility

```diff
@@ -4679,6 +4679,42 @@ template <class _Elem>
 _INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>> =
     _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>;
 
+#ifdef __cpp_lib_concepts
+#ifdef __cpp_lib_byte
+// allow memcmping std::byte with _RANGES equal_to
+template <>
+inline constexpr bool _Equal_memcmp_is_safe_helper<byte, byte, _RANGES equal_to> = true;
+#endif // __cpp_lib_byte
+
+template <class _Pr>
+concept _Equal_to_comparator = is_same_v<_Pr, equal_to<>> || is_same_v<_Pr, _RANGES equal_to>;
+
+// allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to
+// being integral is a precondition of _Value_equality_is_bitwise_equality
+// clang-format off
+template <class _Elem1, class _Elem2, _Equal_to_comparator _Pr>
+    requires (sizeof(_Elem1) == sizeof(_Elem2))
+          && _Is_nonbool_integral<_Elem1> && (!is_volatile_v<_Elem1>)
+          && _Is_nonbool_integral<_Elem2> && (!is_volatile_v<_Elem2>)
+inline constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2, _Pr> =
+    _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
+// clang-format on
+
+template <class _Iter>
+concept _Bitwise_comparable_iterator =
+    (is_pointer_v<_Iter> && !is_volatile_v<remove_pointer_t<_Iter>>) || contiguous_iterator<_Iter>;
+
+// Arbitrary iterators cannot use memcmp
+template <class _Iter1, class _Iter2, class _Pr>
+inline constexpr bool _Equal_memcmp_is_safe = false;
+
+// Determine if it's safe to compare suitable iterators
+template <_Bitwise_comparable_iterator _Iter1, _Bitwise_comparable_iterator _Iter2, class _Pr>
+inline constexpr bool _Equal_memcmp_is_safe<_Iter1, _Iter2, _Pr> =
+    _Equal_memcmp_is_safe_helper<remove_const_t<iter_value_t<_Iter1>>, remove_const_t<iter_value_t<_Iter2>>, _Pr>;

```

Comment:
Since concepts short-circuit, we can simplify this to:
```c++
// Arbitrary iterators cannot use memcmp
template <class _Iter1, class _Iter2, class _Pr>
concept _Equal_memcmp_is_safe = _Bitwise_comparable_iterator<_Iter1> && _Bitwise_comparable_iterator<_Iter2>
    && _Equal_memcmp_is_safe_helper<remove_const_t<iter_value_t<_Iter1>>, remove_const_t<iter_value_t<_Iter2>>;
```
and I observe that the `is_volatile` check in `_Equal_memcmp_is_safe_helper` would be redundant with the check in `_Bitwise_comparable_iterator` if we were passing the reference types of the iterators instead of the value types which are typically cv-unqualified. Why not use:
```suggestion
// Determine if we can compare pointees with memcmp
template <class _Iter1, class _Iter2, class _Pr>
concept _Equal_memcmp_is_safe = contiguous_iterator<_Iter1> && contiguous_iterator<_Iter2>
    && _Equal_memcmp_is_safe_helper<
        remove_const_t<remove_reference_t<iter_reference_t<_Iter1>>>, 
        remove_const_t<remove_reference_t<iter_reference_t<_Iter2>>>, _Pr>;
```
and then get rid of `_Bitwise_comparable_iterator`?

Reply:
I am so happy right now. This is essentially my first try modulo that I screwed up with `Iter_value_t`. The whole `_Bitwise_comparable_iterator` was only to amend that.

Fun fact I already wrote a whiny comment "volatile killed my beautifull code" but thought better of it, do not whine on the internet.

So the message is loud and clear. Always whine about on the internet.